botocore/auth.py

```diff
@@ -245,10 +246,11 @@ def canonical_query_string(self, request):
     def _canonical_query_string_params(self, params):
         # [(key, value), (key2, value2)]
         key_val_pairs = []
-        for key in params:
-            value = str(params[key])
+        if isinstance(params, Mapping):
+            params = params.items()
+        for key, value in params:
             key_val_pairs.append(
-                (quote(key, safe='-_.~'), quote(value, safe='-_.~'))
+                (quote(key, safe='-_.~'), quote(str(value), safe='-_.~'))
```

Comment:
What was the reason for casting the value to str here?

Reply:
I'm not sure, but its currently being cast to a string: https://github.com/boto/botocore/blob/develop/botocore/auth.py#L249
So I didn't want to change that underlying behavior since it wasn't in the scope of this ticket. Do you think I should remove it?

Reply:
Apologies, I read the diff too quickly. That's fine.